### PR TITLE
CARTO: Add QuadbinTileLayer

### PR DESCRIPTION
**Pull request by felixpalmer**
For https://github.com/visgl/deck.gl/issues/6912

#### Background

As an optimization the CARTO backend is switching to a different data format for the Quadkey. This PR implements that change. 

#### Change List
- Add `QuadbinLayer`, `QuadbinTileLayer` and `QuadbinTileset2D` classes
- Modify `CartoLayer` to use `QuadbinTileLayer` rather than `QuadkeyTileLayer`
- Update `carto-dynamic-tile` test app
- Tests

### CODE REVIEW

File: modules/carto/src/layers/carto-layer.ts [hunk 1]
```diff
@@ -304,7 +304,7 @@ export default class CartoLayer<ExtraProps = {}> extends CompositeLayer<
       return [H3TileLayer, props];
     }
     if (data.scheme === 'quadkey') {
```
Comment (alasarr) [inline modules/carto/src/layers/carto-layer.ts:306]: Backend should return the scheme quadbin

File: modules/carto/src/layers/quadbin-tile-layer.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  DefaultProps
+} from '@deck.gl/core';
+import {QuadkeyLayer, QuadkeyLayerProps} from '@deck.gl/geo-layers';
+import QuadbinTileset2D, {quadbinToTile, tileToQuadkey} from './quadbin-tileset-2d';
+import SpatialIndexTileLayer from './spatial-index-tile-layer';
+
+const renderSubLayers = props => {
+  const {data} = props;
+  if (!data || !data.length) return null;
+  return new QuadkeyLayer(props, {
```
Comment (alasarr) [inline modules/carto/src/layers/quadbin-tile-layer.ts:16]: I'm fine keeping the QuadkeyLayer in a first implementation, but as you suggested in the PR description we've room for optimization here

Comment (alasarr) [reply modules/carto/src/layers/quadbin-tile-layer.ts:16]: The main drawback is that the user will receive a Quadkey and they're working with Quadbin, so agree would be better to have the QuadBinLayer and also save some conversions

**Review by alasarr — COMMENTED** [review]
Looks good! Pending things for me, similar to what you pointed out:
- get rid of quadkey layer in favor of quadbin.
- remove all references to quadkey: quadkey-tile-layer and quadkey-tileset-2d

File: modules/carto/src/layers/carto-layer.ts (hunk 1 above)
Comment (charly22) [reply modules/carto/src/layers/carto-layer.ts:306]: Roger, I'll change it to `quadbin` only when `geo_column` is also `quadbin`

Comment (felixpalmer) [reply modules/carto/src/layers/carto-layer.ts]: I've temporarily added support for both

File: modules/carto/src/layers/carto-layer.ts [hunk 3]
```diff
@@ -303,8 +303,8 @@ export default class CartoLayer<ExtraProps = {}> extends CompositeLayer<
     if (data.scheme === 'h3') {
       return [H3TileLayer, props];
     }
-    if (data.scheme === 'quadkey') {
-      return [QuadkeyTileLayer, props];
+    if (data.scheme === 'quadbin' || data.scheme === 'quadkey' /* TODO remove quadkey */) {
```
Comment (alasarr) [inline modules/carto/src/layers/carto-layer.ts:306]: How about remove quadkey before merging?

File: modules/carto/src/index.ts [hunk 4]
```diff
@@ -2,7 +2,8 @@ export {getDefaultCredentials, setDefaultCredentials} from './config';
 export {default as CartoLayer} from './layers/carto-layer';
 export {default as _CartoTileLayer} from './layers/carto-tile-layer';
 export {default as _H3TileLayer} from './layers/h3-tile-layer';
-export {default as _QuadkeyTileLayer} from './layers/quadkey-tile-layer';
+export {default as _QuadbinLayer} from './layers/quadbin-layer';
```
Comment (alasarr) [inline modules/carto/src/index.ts:5]: Do we really need to export _QuadbinLayer?

**Review by alasarr — APPROVED** [review]

File: modules/carto/src/index.ts (hunk 4 above)
Comment (felixpalmer) [reply modules/carto/src/index.ts:5]: @padawannn needs it in Builder I think

File: modules/carto/src/layers/carto-layer.ts (hunk 3 above)
Comment (felixpalmer) [reply modules/carto/src/layers/carto-layer.ts:306]: Yes, agreed. Waiting on deploy from @charly22

Comment (charly22) [reply modules/carto/src/layers/carto-layer.ts:306]: Already deployed

File: modules/carto/src/index.ts (hunk 4 above)
Comment (padawannn) [reply modules/carto/src/index.ts:5]: @alasarr I need it for builder as the rest of the layer exported

Comment (alasarr) [reply modules/carto/src/index.ts:5]: I mean @padawannn needs _QuadbinTileLayer not _QuadbinLayer

Comment (padawannn) [reply modules/carto/src/index.ts:5]: Sorry I really need the `_QuadbinTileLayer`

**Pull request merged**